### DISCUSSION TITLE
Error Index: Intermediate domains don't match

### DISCUSSION THREAD

**Discussion by Shoeboxam** [opendp/opendp#297] [Error Index]
It is easy to make a mistake when chaining that causes an `Intermediate domains don't match` error
or an `Intermediate metrics don't match` error. See the [related section in the user-guide](https://docs.opendp.org/en/latest/user/combinator-constructors.html#chaining-mismatch).

Some common things to check:

1. Domain structure. If there is a structural difference, the error message will include a comparison of the mismatching domains. Review the [Transformation Constructors](https://docs.opendp.org/en/latest/user/transformation-constructors.html) section of the User Guide to make sure you are preprocessing properly.

1. Data types. You may be inadvertently building a transformation that only works with integers- maybe you forgot to add a "`.`" to turn a constructor argument into a float. If you are prescribing custom bit depths in your constructors (you'll know if you are), the chainers are incredibly sensitive to changes between bit depths.

1. Parameters. Some domains carry additional parameters like bounds or size that are not included in the error message. Inspect the parameters used in the surrounding constructors to make sure that the domains are equivalent.

The comments section is open for you to raise any problematic domain/metric mismatch errors you are experiencing.

**Comment by joshua-oss** [comment]
Porting one of the white paper demos over, and I'm unable to determine the appropriate incantations to get a Laplace histogram.

Geometric works fine, like:

```python
geom_histogram = (
    make_split_dataframe(separator=",", col_names=col_names) >>
    make_select_column(key="inc_cat", TOA=str) >>
    make_count_by_categories(categories=income_categories) >>
    make_base_geometric(scale=1.0, D=VectorDomain[AllDomain[int]])
)
```

But Laplace fails with "intermediate domains don't match":

```python
lap_histogram = (
    make_split_dataframe(separator=",", col_names=col_names) >>
    make_select_column(key="inc_cat", TOA=str) >>
    make_count_by_categories(categories=income_categories) >>
    make_cast(TIA=int, TOA=float) >>
    make_base_laplace(scale=1.0, D=VectorDomain[AllDomain[float]])
)
```

The chaining seems to fail on the `make_cast`, and with the `make_cast` removed it gives an obvious type error.

**Reply by Shoeboxam** [reply]
We have a way to specify the output type on `make_count_by_categories` via `TOA` (Atomic Output Type), but have so far excluded floats from the monomorphizations of that constructor. The argument being, "if you have a hardened mechanism that works on integers, then why allow someone to choose a weaker mechanism (the laplace) in the first place?"

I opened an issue with counter-points here:
https://github.com/opendp/opendp/issues/377

And a pull request to allow it here:
https://github.com/opendp/opendp/pull/378

There is example calling code in the integration test in that PR here:
https://github.com/opendp/opendp/blob/e46d4ee20019c1820ac0b28c31c1d2d5a53aba08/python/test/integration/test_histogram.py#L29-L50

EDIT: Patch is in main now.
